rosbag2_cpp/include/rosbag2_cpp/reshapers/stitcher.hpp

```diff
@@ -0,0 +1,156 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// Copyright 2021 Firefly Automatix, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__RESHAPERS__STITCHER_HPP_
+#define ROSBAG2_CPP__RESHAPERS__STITCHER_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+
+// This is necessary because of using stl types here. It is completely safe, because
+// a) the member is not accessible from the outside
+// b) there are no inline functions.
+#ifdef _WIN32
+# pragma warning(push)
+# pragma warning(disable:4251)
+#endif
+
+namespace rosbag2_cpp
+{
+namespace reader_interfaces
+{
+class BaseReaderInterface;
+}  // namespace reader_interfaces
+namespace writer_interfaces
+{
+class BaseWriterInterface;
+}  // namespace writer_interfaces
+
+class ROSBAG2_CPP_PUBLIC Stitcher final
+{
+  static constexpr char const * kDefaultStorageID = "sqlite3";
+
+public:
+  explicit Stitcher(
+    std::unique_ptr<reader_interfaces::BaseReaderInterface> reader_impl =
+    std::make_unique<readers::SequentialReader>(),
+    std::unique_ptr<writer_interfaces::BaseWriterInterface> writer_impl =
+    std::make_unique<writers::SequentialWriter>());
+
+  ~Stitcher();
+
+  /**
+   * Opens the list of URIs and prepares it for stitching.
+   * Each URI in this vector must be a bagfile that exists.
+   * This must be called before any other function is used.
+   *
+   * \note This will step through the directory with the default storage options
+   * * using sqlite3 storage backend
+   * * using no converter options, storing messages with the incoming serialization format
+   * \sa rmw_get_serialization_format.
+   * For specifications, please see \sa open, which let's you specify
+   * more storage and converter options.
+   *
+   * \param storage_uris A vector of URI of the storage to open.
+   * \param output_uri The uri of the storage to write to.
+   **/
+  void open(const std::vector<std::string> & storage_uris, const std::string & output_uri);
```

Comment:
This API is "too low-level" - you should let existing classes do the things they already know how to do. E.g. if you are trying to stitch together "files within a single bag" - then you should point at the _bag_, not the _files_. This should probably be something like

```suggestion
  void open(rosbag2_storage::StorageOptions input_storage, rosbag2_storage::StorageOptions output_storage);
```

This allows you to simply pass `input_storage` to `Reader::open()` (and `output_storage` to `Writer::open()`)

What this API also suggests is that this class is not just a "Stitcher" - it's a "BagReshaper" or a "BagConverter" - what it would actually do is
1. Open an existing bag that was recorded with some configuration
2. Write all messages from that bag to a new bag with some new configuration.

This should allow you to convert from one storage implementation to another one, split large bagfile(s) into more smaller ones, combine many small bagfiles into fewer larger ones. The "Stitch into one" assumption is too limited as an API!

Reply:
ah ok so do you imagine this API to really be doing the "reshaping" and then the "front-end" would be the system that decides what is to be shaped and how? 

In other words, should I rename and redo this from `stitcher` to `reshaper`, with the details of **how** it is being reshaped (stitched, or split) to  the consumer of the API?

Reply:
Yes - I think that this would be more appropriate as a `Reshaper`, and "stitching a multibag into a monobag" is just one possible behavior from given inputs. 

Given your implementation of "read messages one at a time and send to writer" - this more general implementation should look exactly the same under the hood, but provide much more flexible functionality.

Reply:
To go one step further - I think this API is much too large - as far as I can tell given the feature description, the public API should be a single function, like follows:

```
class Rewriter
{
public:
Rewriter(reader, writer);
void rewrite(StorageOptions input_options, StorageOptions output_options);
};
```

Naming things is hard... but I'm almost thinking that, given the description of the functionality, that it's more of a "Rewriter", because it reads a bag, and writes it again with new settings to a new location. It doesn't have to change the shape, it may be changing the storage implementation, compression, or serialization format, without changing the shape.

Reply:
Ok yeah, I think I mimicked the API around the Reader api. But I totally agree it can be stripped down. 

However, Is there not a need to pass in ConverterOptions as well?

Reply:
>  I mimicked the API around the Reader api

It's better to think about what functionality you're trying to provide. Your implementation will call the Reader/Writer APIs, but this thing isn't the same type of thing, so there is no reason for it to have the same API.

> However, Is there not a need to pass in ConverterOptions as well?

Yes, probably

Reply:
See discussion here: https://github.com/ros2/rosbag2/pull/709#discussion_r611096964

I think there is no way to do a many to one bagfile without this API knowing about multiple bags to read.